feat(perf-views) Enable latency drilldown

markstory opened the pull request:
Our distributions and likely those of our customers are not anything approximating normal. Because they are heavily modal, or bi-modal we end up with results that are skewed heavily by outliers. While those outliers are important being able to drill down into the mode is also useful.

Remove the active bar highlighting and instead allow users to zoom in on individual bars to generate new distributions between 0 and the new upper bound.

markstory committed:
feat(perf-views) Enable latency drilldown

Our distributions and likely those of our customers are not anything
approximating normal. Because they are heavily modal, or bi-modal we end
up with results that are skewed heavily by outliers. While those
outliers are important being able to drill down into the mode is also
useful.

Remove the active bar highlighting and instead allow users to zoom in on
individual bars to generate new distributions between 0 and the new
upper bound.

wmak reviewed (approved):
LGTM

markstory commented:
I'm going to see if I can get the 'zoomed in' states to not start the bucketting at 0 as that causes differently shaped mode lumps that aren't ideal.

markstory committed:
Rough in better zooming for histograms

By getting the min value we can get better resolution on data when we
are zoomed in.

markstory committed:
Fix tests.

markstory committed:
Merge branch 'master' into fix/histogram

markstory committed:
Show zoom error in tooltip.

While we want to make sure the API is well behaved we also don't want to
let users get into a bad state.

markstory committed:
Clean up stale code.

markstory commented on src/sentry/snuba/discover.py:
```diff
@@ -187,19 +188,23 @@ def find_histogram_buckets(field, params, conditions):
     if bucket_max == 0:
         raise InvalidSearchQuery(u"Cannot calculate histogram for {}".format(field))
+    bucket_size = ceil((bucket_max - bucket_min) / float(num_buckets))
+    if bucket_size <= 1.0:
```
Capping at 2 might work, capping to 1 gives no useful results because of how the bucketing math works out.

evanh replied:
Why does the math not work? If you have bucket_size of 1, you'll get bars with exact durations.

markstory replied:
You're right. I'm going to figure out why the results were wrong and get that fixed instead of doing this.

markstory committed:
Fix bucket generation with short duration windows.

When the duration conditions create a window equal to or smaller than
the number of bars in the graph we need to include the starting bucket
so that we can zerofill correctly.

markstory committed:
Merge branch 'master' into fix/histogram

markstory committed:
Merge branch 'master' into fix/histogram

markstory committed:
Fix tests.

evanh commented on src/sentry/snuba/discover.py:
```diff
@@ -192,27 +192,24 @@ def find_histogram_buckets(field, params, conditions):
-        raise InvalidSearchQuery(
-            u"Cannot calculate histogram for {}. Resulting buckets are too small.".format(field)
-        )
+    offset = floor(bucket_min / bucket_size)
```
Is there a test for this? I don't understand why the offset wouldn't be equal to the bucket_min. Why do you divide by the bucket size?

markstory committed:
Fix bucket offset calculation
